lighthouse-core/gather/computed/trace-of-tab.js

```diff
@@ -45,18 +45,34 @@ class TraceOfTab extends ComputedArtifact {
       .filter(e => {
         return e.cat.includes('blink.user_timing') ||
           e.cat.includes('loading') ||
-          e.cat.includes('devtools.timeline') ||
-          e.name === 'TracingStartedInPage';
+          e.cat.includes('devtools.timeline');
       })
       // @ts-ignore - stableSort added to Array by WebInspector.
       .stableSort((event0, event1) => event0.ts - event1.ts);
 
-    // The first TracingStartedInPage in the trace is definitely our renderer thread of interest
-    // Beware: the tracingStartedInPage event can appear slightly after a navigationStart
-    const startedInPageEvt = keyEvents.find(e => e.name === 'TracingStartedInPage');
+    // Find out the inspected page frame.
+    /** @type {LH.TraceEvent|undefined} */
+    let startedInPageEvt;
+    const startedInBrowserEvt = keyEvents.find(e => e.name === 'TracingStartedInBrowser');
+    if (startedInBrowserEvt && startedInBrowserEvt.args.data &&
+        startedInBrowserEvt.args.data.persistentIds) {
+      const mainFrame = (startedInBrowserEvt.args.data.frames || []).find(frame => !frame.parent);
+      const mainFrameId = mainFrame && mainFrame.frame;
+      startedInPageEvt = mainFrameId ?
+        Object.assign({}, startedInBrowserEvt, {
+          name: 'TracingStartedInPage', args: {data: {page: mainFrameId}}}) :
+        undefined;
+    }
+    // Support legacy browser versions that do not emit TracingStartedInBrowser event.
+    if (!startedInPageEvt) {
+      // The first TracingStartedInPage in the trace is definitely our renderer thread of interest
+      // Beware: the tracingStartedInPage event can appear slightly after a navigationStart
+      startedInPageEvt = keyEvents.find(e => e.name === 'TracingStartedInPage');
+    }
     if (!startedInPageEvt) throw new LHError(LHError.errors.NO_TRACING_STARTED);
     // @ts-ignore - property chain exists for 'TracingStartedInPage' event.
     const frameId = startedInPageEvt.args.data.page;
```

Comment:
if the casts are annoying, could also do something like 
```js
 /** @type {LH.TraceEvent|undefined} */
let possibleStartedInPageEvt;
...
```
above, and then here
```js
const startedInPageEvt = possibleStartedInPageEvt;
```
because the compiler will know that `possibleStartedInPageEvt` is definitely defined at this point, the type for `startedInPageEvt` will be inferred correctly as `LH.TraceEvent`, and then it won't need any casts below this.

Reply:
I considered that, but I didn't want to introduce more code to workaround compiler issues. I'm fine with casts.